### PR TITLE
[Ops][Feature] Support sequence-parallel MoE at DP=1 for Step3.5/3.7 on Ascend

### PR DESCRIPTION
**Pull request by chenzhh69**
### What this PR does / why we need it?
This PR enables sequence-parallel (SP) MoE support at DP=1 for Step3.5/3.7 models on Ascend by patching the parallel configuration and implementing the SP MoE data flow. Upstream use_sequence_parallel_moe requires dp > 1, so SP is silently disabled for the TP-only (DP=1) layout — even though Ascend's MoE runner self-manages EP communication. 

- is_sequence_parallel threaded into FusedMoEFactory (SP path in MoE runner / shared experts, skip final TP all-reduce).
- SP MoE layers: attention output kept as partial sum (o_proj.reduce_results=False) → pad to TP multiple → reduce-scatter per-rank token shard; dense layers all-gather the full sequence back; residual chunked via sequence_parallel_chunk.
- Step3p5Model.forward all-gathers the final output before logits (MTP draft gathered by the proposer, no model-level gather).

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
CI tests.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Comment by chenzhh69** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Review by wangxiyuan — APPROVED** [review]

File: vllm_ascend/patch/worker/patch_step3p5.py [hunk 1]
```diff
@@ -66,4 +117,325 @@ def _patched_attention_forward(
...
+
+    if self.use_sequence_parallel_moe:
+        tp_world_size = get_tensor_model_parallel_world_size()
+        # small trick using minus, e.g. -17 % 8 = 7
+        sp_pad = (-hidden_states.shape[0]) % tp_world_size
+        # pad if not divisible by world size
+        hidden_states = torch.nn.functional.pad(hidden_states, (0, 0, 0, sp_pad))
+        hidden_states = tensor_model_parallel_reduce_scatter(hidden_states, 0)
+        if not input_is_sequence_parallel:
+            residual = sequence_parallel_chunk(residual)
```
[the remark this answers is not in the document]
Comment (earthmanylf) [reply vllm_ascend/patch/worker/patch_step3p5.py:347]: This mirrors the design of DeepseekV2 SP

File: vllm_ascend/patch/worker/patch_step3p5.py [hunk 2]
```diff
@@ -66,4 +117,325 @@ def _patched_attention_forward(
...
+):
+    # Module-level replacement: zero-arg super() is unavailable outside a
+    # class body, so call the base explicitly.
+    nn.Module.__init__(self)
+
+    self.tp_size = get_tensor_model_parallel_world_size()
+    self.layer_idx = extract_layer_index(prefix)
+    self.is_sequence_parallel = is_sequence_parallel
+
+    self.ep_size = get_ep_group().device_group.size()
```
[the remark this answers is not in the document]
Comment (earthmanylf) [reply vllm_ascend/patch/worker/patch_step3p5.py:127]: This copies the upstream of step3p5's \_\_init\_\_

File: vllm_ascend/patch/worker/patch_step3p5.py [hunk 3]
```diff
@@ -66,4 +117,325 @@ def _patched_attention_forward(
...
+    config = vllm_config.model_config.hf_config
+    quant_config = vllm_config.quant_config
+    parallel_config = vllm_config.parallel_config
+
+    self.hidden_size = config.hidden_size
+    self.enable_eplb = parallel_config.enable_eplb
+    self.n_routed_experts = config.moe_num_experts
+    self.n_logical_experts = self.n_routed_experts
+    self.n_redundant_experts = parallel_config.eplb_config.num_redundant_experts
```
[the remark this answers is not in the document]
Comment (earthmanylf) [reply vllm_ascend/patch/worker/patch_step3p5.py:136]: This copies the upstream of step3p5's \_\_init\_\_

File: vllm_ascend/patch/worker/patch_step3p5.py [hunk 4]
```diff
@@ -66,4 +117,325 @@ def _patched_attention_forward(
...
+    if self.use_fused_all_reduce:
+        logger.warning_once("Enable custom fused all reduce...")
+    else:
+        logger.warning_once("Disable custom fused all reduce...")
+
+    moe_layers_enum = getattr(config, "moe_layers_enum", None)
+    if moe_layers_enum is not None:
+        moe_layers_idx = [int(i) for i in moe_layers_enum.strip().split(",")]
+    else:
+        moe_layers_idx = [i for i in range(1, config.num_hidden_layers)]
```
[the remark this answers is not in the document]
Comment (earthmanylf) [reply vllm_ascend/patch/worker/patch_step3p5.py:275]: This copies the upstream of step3p5's \_\_init\_\_
